DOC, MAINT: Centralize and expand development-related docs (#2734)
intel/scikit-learn-intelex: doc/sources/tests.rst

..
..     http://www.apache.org/licenses/LICENSE-2.0
..
.. Unless required by applicable law or agreed to in writing, software
.. distributed under the License is distributed on an "AS IS" BASIS,
.. WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
.. See the License for the specific language governing permissions and
.. limitations under the License.
.. include:: substitutions.rst

=============
Running Tests
=============

Overview
--------

The |sklearnex| contains a test suite consisting of a mixture of smoke tests around patching along with unit tests, which are written in a mixture of Python's ``unittest`` (for legacy interfaces) and ``pytest``. However, all of the tests are executed with ``pytest`` as runner. Apart from the tests, code examples are also executed, but are not thoroughly checked for correctness, just for executing without erroring out.

Running test scripts
--------------------

Requirements
~~~~~~~~~~~~

As the library is designed with optional components and integrates with external packages that are optional by design, executing the tests involves additional dependencies, some of which are mandatory and some of which are optional.

> The mandatory dependencies for tests with locked versions of packages are listed under a file `requirements-test.txt <https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/requirements-test.txt>`__, but it is not strictly necessary to have the exact versions listed there. Those locked requirements can be installed with ``pip`` as follows:

Reviewer: ```suggestion
The mandatory dependencies for tests with locked versions of packages are listed under a file `requirements-test.txt <https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/requirements-test.txt>`__, but it is not strictly necessary to have the exact versions listed there. Those versioned requirements can be installed with ``pip`` as follows:
```

.. code-block:: bash

    pip install -r requirements-test.txt

Some tests will only execute depending on the availability of optional dependencies at runtime. Other optional dependencies that will trigger additional tests can be installed as follows, assuming a Linux* system:

.. code-block:: bash

    pip install \
        dpctl `# for GPU functionalities` \
        dpnp `# for array API and GPU functionalities`

    pip install --index-url https://software.repos.intel.com/python/pypi \
        torch `# for array API`

    pip install --index-url https://software.repos.intel.com/python/pypi \
        mpi4py impi_rt `# for distributed mode, be sure to install from Intel's index` \
        && pip install pytest-mpi `# also required, but not from Intel's index`

> .. warning:: It might not be possible to install all of the test dependencies simultaneously in the same Python environment. In particular, dependencies ``torch`` and ``dpctl`` / ``dpnp`` are likely to not be installable in a compatible way in the same environment if using pre-built distributions. Try using different Python environments for each set of dependencies to test.

Reviewer: ```suggestion
.. warning:: It might not be possible to install all of the test dependencies simultaneously in the same Python environment. In particular, dependencies of ``torch`` and ``dpctl`` / ``dpnp`` are potentially incompatible if using pre-built distributions. Try using different Python environments for each set of dependencies to test.
```

Executing tests
~~~~~~~~~~~~~~~

In order to run the whole test suite, the following script can be used on Linux*:

.. code-block:: bash

    conda-recipe/run_test.sh

.. warning:: This script must be executed from the root of the repository.

It also comes with an analog for Windows*:

.. code-block:: console

    call conda-recipe/run_test.bat

Individual test files or tests can be executed with PyTest under different options (such as a different verbosity mode, stopping at the first failure, etc.) with calls to pytest - for example:

.. code-block:: bash

    pytest sklearnex/ensemble/tests/test_forest.py

.. hint:: If executing these from the root of the repository, there might be naming clashes between the folders and the installed Python modules. It might be helpful to :ref:`build the library extensions in-place <build_script>` and set ``$PYTHONPATH`` to avoid problems.

Configurable options
~~~~~~~~~~~~~~~~~~~~

The files ``run_test.sh`` and ``run_test.bat`` offer configurable behaviors through environment variables and command line arguments:

> - Environment variable ``$NO_DIST``, if set, will make it avoid running distributed mode tests. Note that executing these tests requires additional dependencies, otherwise they will be skipped either way.

Reviewer: ```suggestion
- Environment variable ``$NO_DIST``, if set, will avoid running distributed mode tests. Note that executing these tests requires additional dependencies, otherwise they will be skipped either way.
```

.. warning:: This script needs to be executed from the root of the repository as working directory. The script will modify the working directory when it launches subprocesses, so if using environment variables like ``$PYTHONPATH``, these need to be set as absolute paths (not relative).

.. _conformance_tests:

Scikit-learn's test suite
-------------------------

The |sklearnex| is regularly tested for correctness through the test suite of |sklearn| itself executed with patching applied, referred throughout the CI jobs and files as 'conformance testing'.

Executing tests
~~~~~~~~~~~~~~~

To execute the |sklearn| conformance tests, the following script can be used:

.. code-block:: bash

    ./.ci/scripts/run_sklearn_tests.sh

Note that some tests are known to produce failures - for example, :obj:`sklearn.linear_model.LinearRegression` allows an argument ``copy_X``, and one of their tests checks that passing ``copy_X=False`` modifies the 'X' input in-place, while the |sklearnex| never modifies this data regardless of the argument ``copy_X``, hence the test would show a failure under a patched call to |sklearn|, even though the results do not change.

> Cases that are known to fail are not executed during these conformance test. The list of deselected tests can be found under `deselected_tests.yaml <https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/deselected_tests.yaml>`__.

Reviewer: ```suggestion
Cases that are known to fail are not executed during these conformance tests. The list of deselected tests can be found under `deselected_tests.yaml <https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/deselected_tests.yaml>`__.
```
